common/src/main/java/com/wynntils/core/text/StyledText.java

```diff
@@ -0,0 +1,487 @@
+/*
+ * Copyright © Wynntils 2023.
+ * This file is released under AGPLv3. See LICENSE for full license details.
+ */
+package com.wynntils.core.text;
+
+import com.google.common.collect.Iterables;
+import com.wynntils.utils.mc.ComponentUtils;
+import com.wynntils.utils.type.Pair;
+import java.util.Collections;
+import java.util.Deque;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Objects;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+import net.minecraft.ChatFormatting;
+import net.minecraft.network.chat.ClickEvent;
+import net.minecraft.network.chat.Component;
+import net.minecraft.network.chat.HoverEvent;
+import net.minecraft.network.chat.MutableComponent;
+import net.minecraft.network.chat.Style;
+import org.apache.commons.lang3.ArrayUtils;
+
+public final class StyledText {
+    private final Component temporaryWorkaround;
+
+    private final List<StyledTextPart> parts;
+
+    private final List<ClickEvent> clickEvents;
+    private final List<HoverEvent> hoverEvents;
+
+    StyledText(Component component) {
+        temporaryWorkaround = component;
+
+        parts = new LinkedList<>();
+        clickEvents = new LinkedList<>();
+        hoverEvents = new LinkedList<>();
+
+        // Walk the component tree using DFS
+        // Save the style of the parent component so we can inherit it
+        Deque<Pair<Component, Style>> deque = new LinkedList<>();
+
+        deque.add(new Pair<>(component, Style.EMPTY));
+
+        while (!deque.isEmpty()) {
+            Pair<Component, Style> currentPair = deque.pop();
+            Component current = currentPair.key();
+            Style parentStyle = currentPair.value();
+
+            String componentString =
```

Comment:
Ok, this is wrong like I suspected but could not really say for sure before. What you really need to do here is basically call `StyledText.fromString(current.getString())`. Since this individual component element can have a style like RED, and a string like "warning: §kDANGER§r ahead", and in such a case, it will need to be replaced with not one, but three StyledTextParts.

Reply:
(And prectically, you probably want to solve this by an additional static method like `getStyledTextParts(String codedString)`, which just gives you a List<StyledTextParts>, and that you can use both here and in `StyledText.fromString`.

Reply:
Yes, I understand that. You are getting THIS component's string. And now let's assume that this string is "warning: §kDANGER§r ahead". Then we need to create *three* StyledTextParts from it, right?

Reply:
I see what you mean now.. Yes, most likely..

Reply:
I handled this, and also merged StyledText fromString with the StyledTextPart#fromString. I know the location is semi-ideal for the method, but this is the most sane way, with the least code duplication.